django/utils/dateformat.py

```diff
@@ -24,24 +24,25 @@
 )
 from django.utils.translation import gettext as _
 
-re_formatchars = _lazy_re_compile(r'(?<!\\)([aAbcdDeEfFgGhHiIjlLmMnNoOPrsStTUuwWyYzZ])')
+re_formatchars = _lazy_re_compile(r'((?:^|(?<!\\))(?:\\\\)*)(\\*)([aAbcdDeEfFgGhHiIjlLmMnNoOPrsStTUuwWyYzZ])')
 re_escaped = _lazy_re_compile(r'\\(.)')
 
 
 class Formatter:
+
+    def _replace_token(self, match):
+        prefix, escape, specifier = match.groups()
+        if escape:
+            return prefix + specifier
+        if not isinstance(self.data, (datetime.datetime, datetime.time)) and hasattr(TimeFormat, specifier):
+            raise TypeError(
+                'The format for date objects may not contain time-related '
+                f'format specifiers (found {specifier!r}).'
+            )
```

Comment:
This will be called for every token and is inefficient. Perhaps we should just check once upfront whether it is a `datetime.date` object and pass a different, more limited regex? Do we have to handhold?

Currently there is also nothing to stop a `datetime.time` instance being passed to `DateFormat.format()` either, so why this? Perhaps we could improve this? I'm not sure we really need a separate `DateFormat` and `TimeFormat` if we just `isinstance()` check before and use a different regex? Or if we want to keep this message we could have a faster implementation that has a static set of characters specific to the type of object to check against?

ISTM that we should have quite a bit of flexibility as none of this is documented API.